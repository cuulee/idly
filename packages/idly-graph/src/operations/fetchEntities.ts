import { ImMap, ImSet } from 'idly-common/lib/misc/immutable';
import { EntityId } from 'idly-common/lib/osm/structures';

import { Tree } from '../graph/Tree';
import { channelBuilder } from '../misc/channelBuilder';
import { Manager } from '../worker/store/manager';
import { WorkerStateAccessActions } from './types';

export interface WorkerGetEntities {
  readonly type: WorkerStateAccessActions.FetchEntities;
  readonly request: {
    readonly entityIds: EntityId[];
  };
}

export type ReturnType = Tree;

/** Main Thread */

export function fetchEntities(
  connector: any,
): (req: WorkerGetEntities['request']) => Promise<ReturnType> {
  const channel = channelBuilder<WorkerGetEntities>(connector)(
    WorkerStateAccessActions.FetchEntities,
  );
  return async request => {
    const json = await channel(request);
    const parsedTree: ReturnType = Tree.fromString(json);
    return parsedTree;
  };
}

/** Worker Thread */

export function workerFetchEntities(
  manager: Manager,
): (request: WorkerGetEntities['request']) => Promise<string> {
  return async ({ entityIds }) => {
    const entities = await manager.entityLookup(entityIds).map(e => [e.id, e]);
    const toReturn: ReturnType = Tree.fromObject({
      deletedIds: ImSet(),
      entityTable: ImMap(entities),
      knownIds: ImSet(entityIds),
    });
    return JSON.stringify(toReturn.toJs());
  };
}
