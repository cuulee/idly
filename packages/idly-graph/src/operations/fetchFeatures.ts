import { Feature } from 'idly-common/lib/osm/feature';
import { EntityId } from 'idly-common/lib/osm/structures';

import { channelBuilder } from '../misc/channelBuilder';
import { Manager } from '../worker/store/manager';
import { WorkerActions } from './types';

export interface WorkerFetchFeatures {
  readonly type: WorkerActions.FetchFeatures;
  readonly request: {
    readonly entityIds: EntityId[];
  };
}

export type ReturnType = Array<Feature<any, any>>;

/** Main Thread */

export function fetchFeatures(
  connector: any,
): (req: WorkerFetchFeatures['request']) => Promise<ReturnType> {
  const channel = channelBuilder<WorkerFetchFeatures>(connector)(
    WorkerActions.FetchFeatures,
  );
  return async request => {
    const json = await channel(request);
    const parsedFeatures: ReturnType = JSON.parse(json);
    return parsedFeatures;
  };
}

/** Worker Thread */

export function workerFetchFeatures(
  manager: Manager,
): (request: WorkerFetchFeatures['request']) => Promise<string> {
  return async ({ entityIds }) => {
    const toReturn: ReturnType = await manager.featureLookup(entityIds);
    console.log(entityIds, toReturn);
    return JSON.stringify(toReturn);
  };
}
