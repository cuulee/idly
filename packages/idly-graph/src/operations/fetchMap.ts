import { FeatureCollection, featureCollection } from '@turf/helpers';
import { BBox } from 'idly-common/lib/geo/bbox';

import { channelBuilder } from '../misc/channelBuilder';
import { Manager } from '../worker/store/manager';
import { WorkerActions } from './types';

export interface WorkerFetchMap {
  readonly type: WorkerActions.FetchMap;
  readonly request: {
    readonly bbox: BBox;
    readonly zoom: number;
  };
}
/**
 * this type helps in keeping the parsing worker / main side in
 * sync.
 */
export type ReturnType = GeoJSON.FeatureCollection<any>;

/** Main Thread */
/**
 *
 * @param connector
 */
export function fetchMap(
  connector: any,
): (req: WorkerFetchMap['request']) => Promise<ReturnType> {
  const channel = channelBuilder<WorkerFetchMap>(connector)(
    WorkerActions.FetchMap,
  );
  return async req => {
    const json = await channel(req);
    const fc: ReturnType = JSON.parse(json);
    return fc;
  };
}

/** Worker Thread */

export function workerFetchMap(
  manager: Manager,
): (request: WorkerFetchMap['request']) => Promise<string> {
  return async ({ bbox, zoom }) => {
    const r = await manager.receive(bbox, zoom);
    const toReturn: ReturnType = featureCollection(r);
    return JSON.stringify(toReturn);
  };
}
