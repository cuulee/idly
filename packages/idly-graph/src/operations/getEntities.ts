import { ImMap, ImSet } from 'idly-common/lib/misc/immutable';
import { EntityId } from 'idly-common/lib/osm/structures';

import { Tree } from '../graph/Tree';
import { getChannelBuilder } from '../misc/channelBuilder';
import { recursiveLookup } from '../misc/recursiveLookup';
import {
  GetActions,
  Operation,
  WorkerOperation,
  WorkerState,
} from './operationsTypes';

export interface GetEntities {
  readonly type: GetActions.GetEntities;
  readonly request: {
    readonly entityIds: EntityId[];
  };
  readonly response: Tree;
}

/** Main Thread */
export function getEntities(connector: any): Operation<GetEntities> {
  const channel = getChannelBuilder<GetEntities>(connector)(
    GetActions.GetEntities,
  );
  return async request => {
    const json = await channel(request);
    const parsedTree: GetEntities['response'] = Tree.fromString(json);
    return parsedTree;
  };
}

/** Worker Thread */
export function workerGetEntities(
  state: WorkerState,
): WorkerOperation<GetEntities> {
  return async ({ entityIds }) => {
    const entities = entityIds
      .map(id => recursiveLookup(id, state.entityTable))
      .reduce((prev, curr) => prev.concat(curr), [])
      .map(e => [e.id, e]);
    const toReturn: GetEntities['response'] = Tree.fromObject({
      deletedIds: ImSet(),
      entityTable: ImMap(entities),
      knownIds: ImSet(entityIds),
    });
    return toReturn.toString();
  };
}

// function workerGetEntitiesTest(
//   state: WorkerState,
// ): (t: GetEntities['request']) => Promise<GetEntities['response']> {
//   return async ({ entityIds }) => {
//     const entities = entityIds
//       .map(id => recursiveLookup(id, state.entityTable))
//       .reduce((prev, curr) => prev.concat(curr), [])
//       .map(e => [e.id, e]);
//     return Tree.fromObject({
//       deletedIds: ImSet(),
//       entityTable: ImMap(entities),
//       knownIds: ImSet(entityIds),
//     });
//   };
// }

// const request = {
//   async serializer(input: GetEntities['request']): Promise<string> {
//     return JSON.stringify(input);
//   },
//   async deserializer(string: string): Promise<GetEntities['request']> {
//     return JSON.parse(string);
//   },
//   async logic(input: GetEntities['request']): Promise<GetEntities['response']> {
//     return Tree.fromEntities([]);
//   },
// };

// export interface Smartie<T> {
//   readonly type: string;
//   readonly request;
//   readonly response: Tree;
// }

// function connectToWorker(
//   connector: any,
//   type: any,
//   request: any,
//   serializer: any,
//   deserializer: any,
// ): any {
//   const toSend = {
//     request,
//     type,
//   };
//   return connector.postMessage(toSend).catch((e: Error) => {
//     // tslint:disable-next-line:no-expression-statement
//     console.log('Worker Error', e.message);
//     return Promise.reject(e.message);
//   });
// }

// function connectToMain(record: Map<string, any>) {
//   return (type: string, cb: any) => {
//     // tslint:disable-next-line:no-expression-statement
//     record.set(type, cb);
//     return record;
//   };
// }

// const response = {
//   async serializer(input: GetEntities['response']): Promise<string> {
//     return JSON.stringify(input.toJs());
//   },
//   async deserializer(string: string): Promise<GetEntities['response']> {
//     return Tree.fromString(string);
//   },
//   async logic(
//     state: any,
//     { entityIds }: GetEntities['request'],
//   ): Promise<GetEntities['response']> {
//     const entities = entityIds
//       .map(id => recursiveLookup(id, state.entityTable))
//       .reduce((prev, curr) => prev.concat(curr), [])
//       .map(e => [e.id, e]);
//     return Tree.fromObject({
//       deletedIds: ImSet(),
//       entityTable: ImMap(entities),
//       knownIds: ImSet(entityIds),
//     });
//   },
// };
