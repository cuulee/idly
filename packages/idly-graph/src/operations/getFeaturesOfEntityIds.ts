import { Feature, LineString, Point, Polygon } from '@turf/helpers';
import { entityTableGen } from 'idly-common/lib/osm/entityTableGen';
import { Entity, EntityId, EntityTable } from 'idly-common/lib/osm/structures';

import { getChannelBuilder } from '../misc/channelBuilder';
import { recursiveLookup } from '../misc/recursiveLookup';
import { entityToFeature } from '../thread/entityToFeatures';
import {
  GetActions,
  Operation,
  WorkerOperation,
  WorkerState,
} from './operationsTypes';

export interface GetFeaturesOfEntityIds {
  readonly type: GetActions.GetFeaturesOfEntityIds;
  readonly request: {
    readonly entityIds: EntityId[];
  };
  readonly response: Array<Feature<Point | Polygon | LineString>>;
}

export function getFeaturesOfEntityIds(
  connector: any,
): Operation<GetFeaturesOfEntityIds> {
  const channel = getChannelBuilder<GetFeaturesOfEntityIds>(connector)(
    GetActions.GetFeaturesOfEntityIds,
  );
  return async request => {
    const json = await channel(request);
    const parsedFeatures: GetFeaturesOfEntityIds['response'] = JSON.parse(json);
    return parsedFeatures;
  };
}

/** Worker Thread */

export function workerGetFeaturesOfEntityIds(
  state: WorkerState,
): WorkerOperation<GetFeaturesOfEntityIds> {
  return async ({ entityIds }) => {
    const entities = entityIds
      .map(id => recursiveLookup(id, state.entityTable))
      .reduce((prev: Set<Entity>, curr) => {
        for (const e of curr) {
          // tslint:disable-next-line:no-expression-statement
          prev.add(e);
        }
        return prev;
      }, new Set<Entity>());

    const entityTable: EntityTable = entityTableGen(entities);
    const workerPlugins = await state.plugins;
    const toReturn: GetFeaturesOfEntityIds['response'] = entityToFeature(
      workerPlugins.map((r: any) => r.worker),
    )(entityTable);
    return JSON.stringify(toReturn);
  };
}
