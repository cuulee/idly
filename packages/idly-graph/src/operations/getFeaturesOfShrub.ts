import { Feature, LineString, Point, Polygon } from '@turf/helpers';
import { Shrub } from '../graph/Shrub';

import { Tree } from '../graph/Tree';
import { getChannelBuilder } from '../misc/channelBuilder';
import { entityToFeature } from '../thread/entityToFeatures';
import {
  GetActions,
  Operation,
  WorkerOperation,
  WorkerState,
} from './operationsTypes';

export interface GetFeaturesOfShrub {
  readonly type: GetActions.getFeaturesOfShrub;
  readonly request: {
    readonly shrubString: string;
  };
  readonly response: Array<Feature<Point | LineString | Polygon>>;
}

export function getFeaturesOfShrub(
  connector: any,
): Operation<GetFeaturesOfShrub> {
  const channel = getChannelBuilder<GetFeaturesOfShrub>(connector)(
    GetActions.getFeaturesOfShrub,
  );
  return async request => {
    const json = await channel(request);
    const parsedFeatures: GetFeaturesOfShrub['response'] = JSON.parse(json);
    return parsedFeatures;
  };
}

/** Worker Thread */

export function workergetFeaturesOfShrub(
  state: WorkerState,
): WorkerOperation<GetFeaturesOfShrub> {
  return async ({ shrubString }) => {
    const shrub = Shrub.fromString(shrubString);
    const { entityTable, knownIds } = shrub.toObject();
    // TODO figure out deleted entities, for now I will subtract them
    //  from entityTable :?
    let filteredEntityTable = entityTable;
    knownIds.forEach(id => {
      filteredEntityTable = filteredEntityTable.delete(id);
    });

    const workerPlugins = await state.plugins;
    const toReturn: GetFeaturesOfShrub['response'] = entityToFeature(
      workerPlugins.map((r: any) => r.worker),
    )(filteredEntityTable);
    return JSON.stringify(toReturn);
  };
}
