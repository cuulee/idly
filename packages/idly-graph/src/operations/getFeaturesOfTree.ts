import { entityTableGen } from 'idly-common/lib/osm/entityTableGen';
import { Feature } from 'idly-common/lib/osm/feature';
import { EntityId, EntityTable } from 'idly-common/lib/osm/structures';

import { Tree } from '../graph/Tree';
import { getChannelBuilder } from '../misc/channelBuilder';
import { recursiveLookup } from '../misc/recursiveLookup';
import { entityToFeature } from '../thread/entityToFeatures';
import {
  GetActions,
  Operation,
  WorkerOperation,
  WorkerState,
} from './operationsTypes';

export interface GetFeaturesOfTree {
  readonly type: GetActions.GetFeaturesOfTree;
  readonly request: {
    readonly treeString: string;
  };
  readonly response: Array<Feature<any, any>>;
}

export function getFeaturesOfTree(
  connector: any,
): Operation<GetFeaturesOfTree> {
  const channel = getChannelBuilder<GetFeaturesOfTree>(connector)(
    GetActions.GetFeaturesOfTree,
  );
  return async request => {
    const json = await channel(request);
    const parsedFeatures: GetFeaturesOfTree['response'] = JSON.parse(json);
    return parsedFeatures;
  };
}

/** Worker Thread */

export function workerGetFeaturesOfTree(
  state: WorkerState,
): WorkerOperation<GetFeaturesOfTree> {
  return async ({ treeString }) => {
    const tree = Tree.fromString(treeString);
    const { entityTable, deletedIds } = tree.toObject();
    // TODO figure out deleted entities, for now I will subtract them
    //  from entityTable :?
    const filteredTable = entityTable
      .filter((v, k = '') => {
        return !deletedIds.has(k);
      })
      .toMap();
    const workerPlugins = await state.plugins;
    const toReturn: GetFeaturesOfTree['response'] = entityToFeature(
      workerPlugins.map((r: any) => r.worker),
    )(filteredTable);
    return JSON.stringify(toReturn);
  };
}
