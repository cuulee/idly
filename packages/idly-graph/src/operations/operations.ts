import { Manager } from '../worker/store/manager';
import { workerFetchEntities } from './fetchEntities';
import { workerFetchMap } from './fetchMap';
import { WorkerActions, WorkerActionsType } from './types';

function controller(
  manager: Manager,
  message: WorkerActionsType,
): Promise<string> {
  switch (message.type) {
    case WorkerActions.FetchMap:
      return workerFetchMap(manager)(message.request);
    case WorkerActions.FetchEntities:
      return workerFetchEntities(manager)(message.request);
    default: {
      console.error('no handler for', message.type);
      return Promise.resolve(message.type);
    }
  }
}

/**
 * wraps the Promise<string> to the shape of `WorkerResponse`,
 * this helps in sending errors to main thread and also prevents
 * promise works JSON.parse and lets you handle custom JSON parsing
 * of certain objects like immutable, Tree etc.
 * @param message
 */
export function operations(
  plugins: Promise<any>,
): (message: WorkerActionsType) => Promise<any> {
  const manager = new Manager(plugins);
  return message => {
    return controller(manager, message);
  };
}
