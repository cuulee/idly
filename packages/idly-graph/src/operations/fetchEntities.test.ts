import { pluginsStub } from '../misc/pluginsStub';
import { PromiseWorkerStub } from '../misc/PromiseWorkerStub';
import { miniXML3 } from '../worker/parsing/fixtures';
import { fetchEntities } from './fetchEntities';
import { fetchMap } from './fetchMap';
import { operations } from './operations';

declare var global: any;
// tslint:disable no-expression-statement no-object-mutation

describe('fetchEntities', () => {
  test('small xml test', async () => {
    global.fetch = jest.fn().mockImplementation(() => {
      return new Promise((resolve, reject) => {
        resolve({
          id: '123',
          text(): Promise<any> {
            return Promise.resolve(miniXML3);
          },
          ok: true,
        });
      });
    });
    const promiseWorker = new PromiseWorkerStub();
    const controller = operations(pluginsStub());
    promiseWorker.registerPromiseWorker(controller);
    await fetchMap(promiseWorker)({
      bbox: [
        -73.98630242339283,
        40.73537277780156,
        -73.98264244865518,
        40.73941515574535,
      ],
      zoom: 17.54,
    });
    const resp1 = fetchEntities(promiseWorker)({ entityIds: ['n1'] });
    const resp2 = fetchEntities(promiseWorker)({
      entityIds: ['n2', 'w1'],
    });
    expect(await resp1).toMatchSnapshot();
    expect(await resp2).toMatchSnapshot();
  });
});
