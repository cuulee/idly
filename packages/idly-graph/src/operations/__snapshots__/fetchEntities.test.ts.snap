// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`fetchEntities small xml test 1`] = `
Tree {
  "_deletedIds": Immutable.Set [],
  "_entityTable": Immutable.Map {
    "n1": Object {
      "attributes": Object {
        "changeset": "38701677",
        "timestamp": "2016-04-19T17:58:40Z",
        "uid": "592097",
        "user": "Dombo",
        "version": "1",
        "visible": true,
      },
      "id": "n1",
      "loc": Object {
        "lat": 34.530396,
        "lon": 74.1816382,
      },
      "tags": "[]",
      "type": "node",
    },
  },
  "_knownIds": Immutable.Set [
    "n1",
  ],
  "_parentWays": Immutable.Map {},
}
`;

exports[`fetchEntities small xml test 2`] = `
Tree {
  "_deletedIds": Immutable.Set [],
  "_entityTable": Immutable.Map {
    "n2": Object {
      "attributes": Object {
        "changeset": "38701677",
        "timestamp": "2016-04-19T17:58:40Z",
        "uid": "592097",
        "user": "Dombo",
        "version": "1",
        "visible": true,
      },
      "id": "n2",
      "loc": Object {
        "lat": 34.530396,
        "lon": 74.1816382,
      },
      "tags": "[]",
      "type": "node",
    },
    "n1": Object {
      "attributes": Object {
        "changeset": "38701677",
        "timestamp": "2016-04-19T17:58:40Z",
        "uid": "592097",
        "user": "Dombo",
        "version": "1",
        "visible": true,
      },
      "id": "n1",
      "loc": Object {
        "lat": 34.530396,
        "lon": 74.1816382,
      },
      "tags": "[]",
      "type": "node",
    },
    "w1": Object {
      "attributes": Object {
        "changeset": "50089927",
        "timestamp": "2017-07-06T17:19:04Z",
        "uid": "2815653",
        "user": "Srihari Thalla",
        "version": "2",
        "visible": true,
      },
      "id": "w1",
      "nodes": Array [
        "n1",
        "n2",
      ],
      "tags": "[]",
      "type": "way",
    },
  },
  "_knownIds": Immutable.Set [
    "n2",
    "w1",
  ],
  "_parentWays": Immutable.Map {
    "n1": Immutable.Set [
      "w1",
    ],
    "n2": Immutable.Set [
      "w1",
    ],
  },
}
`;
